osu.Server.Queues.ScoreStatisticsProcessor/Helpers/WebRequestHelper.cs

```diff
@@ -0,0 +1,222 @@
+                }
+
+                throw;
+            }
+        }
+
+        private static string hmacEncode(string input, byte[] key)
+        {
+            byte[] byteArray = Encoding.ASCII.GetBytes(input);
+
+            using (var myhmacsha1 = new HMACSHA1(key))
+            {
+                byte[] hashArray = myhmacsha1.ComputeHash(byteArray);
+                return hashArray.Aggregate(string.Empty, (s, e) => s + $"{e:x2}", s => s);
+            }
+        }
+
+        /// <remarks>
+        /// <para>
+        /// Anyone reading this without context is probably thinking "what sort of abomination is this? isn't this all wrong???"
+        /// The answer is that this is certainly weird, some may call it an abomination, but it <i>should</i> - given a specific set of circumstances that happens to be true - work.
+        /// </para>
+        /// <para>
+        /// The simple way to retrieve a score's rank in the beatmap leaderboard would be to use the database.
+        /// That cannot be done because the database is slow (>30s lookup times).
+        /// The database cannot really be made any faster, it has a bunch of indexes on it already.
+        /// </para>
+        /// <para>
+        /// Therefore, we want to leverage elasticsearch in some way.
+        /// However, this score at the time this method is going to be called, <i>will not be</i> in elasticsearch, as that happens <i>after</i> <c>osu-queue-score-statistics</c> processing is done.
+        /// That said, you don't actually <i>need</i> the score to be in ES at this point in time to retrieve its global rank in the beatmap leaderboard.
+        /// All you need to determine <i>that</i> is the score's total and its ID, as those are the two determining factors of how many scores precede this one in the beatmap leaderboard.
+        /// Add a 1 to that number and you get this score's rank.
+        /// </para>
+        /// <para>
+        /// At this point there are two options: either hit ES directly from here, or hit <c>osu-web</c> which already does the same trick to retrieve global rank of a score in the beatmap leaderboard.
+        /// The first approach was deemed undesirable, because it is preferred that there is a single connection point to ES in the infrastructure.
+        /// Therefore, the <c>GET /scores/{id}</c> API endpoint should do the trick for the second approach.
+        /// If this ever becomes a problem the proposition is to move the ES lookup logic to https://github.com/ppy/osu-global-rank-lookup-cache and have both web and this call into that component.
+        /// </para>
+        /// <para>
+        /// Note that this working correctly also relies on the fact that <c>osu-web</c> API has lazer mode turned forcefully on with no way to toggle it off
+        /// (see https://github.com/ppy/osu-web/blob/f46806bb81eb0d3b0807c70bf8e3dc13f0783ec9/app/Libraries/Search/ScoreSearchParams.php#L86-L88).
+        /// </para>
+        /// </remarks>

```

Comment:
I'd like anyone reviewing this to read *all of this* over *carefully* and double-check all of the claims made here because this is a pretty precarious setup.

Reply:
This all reads correctly to me. Reading over this, the only real concern I would have is with web API not being able to handle the number of requests it receives, but since there's no large JSON decode involved and it's being limited to user high scores already, I think we should be okay.